src/utils/components/EnvironmentEditor/hooks/useEnvironments.ts

```diff
@@ -30,72 +32,89 @@ type UseEnvironmentsType = {
 
 const useEnvironments = (
   vm: V1VirtualMachine,
-  updateVM: (updatedVM: V1VirtualMachine) => Promise<V1VirtualMachine | void>,
+  originalVM: V1VirtualMachine,
+  updateVM: Updater<V1VirtualMachine>,
   onEditChange?: (edited: boolean) => void,
 ): UseEnvironmentsType => {
   const { t } = useKubevirtTranslation();
-  const initialEnvironments = React.useMemo(() => getVMEnvironmentsVariables(vm), [vm]);
-  const [environments, setEnvironments] = React.useState(initialEnvironments);
-  const [environmentsEdited, setEnvironmentsEdited] = React.useState(false);
-  const [error, setError] = React.useState<Error>();
-
-  const onEnvironmentAdd = React.useCallback(() => {
-    setEnvironments((envs) =>
-      envs.concat([{ name: undefined, serial: getRandomSerial().toUpperCase(), kind: undefined }]),
-    );
-  }, []);
-
-  const onEnvironmentChange = React.useCallback(
-    (environmentIndex: number, name: string, serial: string, kind: EnvironmentKind) => {
-      if (environments.find((env, index) => env.name === name && index !== environmentIndex)) {
-        return setError(new Error(t('Resource already selected')));
+  const [error, setError] = useState<Error>();
+  const [edited, setEdited] = useState(false);
+  const originalEnvironments = useMemo(() => getVMEnvironmentsVariables(originalVM), [originalVM]);
+  const environments = useMemo(() => getVMEnvironmentsVariables(vm), [vm]);
+
+  useEffect(() => {
+    const envsEdited = areEnvironmentsChanged(environments, originalEnvironments);
+
+    if (envsEdited !== edited) {
+      setEdited(envsEdited);
+      if (onEditChange) onEditChange(envsEdited);
+    }
+  }, [edited, environments, onEditChange, originalEnvironments]);
+
+  const onEnvironmentAdd = useCallback(() => {
+    updateVM((draftVM: V1VirtualMachine) => {
+      if (!draftVM.spec.template?.spec?.domain?.devices?.disks) {
```

Comment:
just ensure to have this array in the add. In change and remove, at least one disk exist ( the environment disk )/